### ISSUE TITLE
package-lock.json not updating "resolved" field after moving registry

### ISSUE THREAD

**Issue by xenobytezero** [npm/npm#19578]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [x] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

We have recently moved from using a NPM proxy as our NPM registry (via Artifactory) to just using the default registry and scoping our custom packages and getting them from Artifactory.

After making the change and removing the line in the `.npmrc` file that pointed everything at Artifactory, everything has successfully installed however our `package-lock.json` file still has it's `resolved` field for each public package pointing at our Artifactory URL.

I don't want to delete our `package-lock.json` and recreate it, first because I need to keep all our versions locked, and recreating it changes all our transitive dependencies, and second cause I would like to prove that the change to the public repo does not change any of the checksums for the packages, proving (as close as we can get) zero impact to the actual project. I also don't want to manually change the URLs of the `package-lock.json` as it becomes more difficult to make the case for no package changes.

#### How can the CLI team reproduce the problem?

1. Point NPM at a private registry (using a local `.npmrc` file with a `registry=<private registry>` line)
2. Run `npm install`
3. Note the `resolved` field of the generated `package-lock.json`
4. Remove the line added in 1.
5. Run `npm install`
6. Note the `resolved` field of the generated `package-lock.json` still points at the private registry

### supporting information:

```
>npm -v
5.5.1
>node -v
v8.9.2

Windows 10x64
```

**Comment by mmisztal1980** [comment]
I'm seeing this as well. 

Basically I've had a private registry setup in my .npmrc and now I've added a second one. I've set the active registry to be the later one, and after deleting **package.lock.json** and `npm install` I'm seeing package.lock.json contain 'resolved' properties pointing to the old registry.

Can anyone advise?

**Comment by babeal** [comment]
I'm seeing the opposite.  I have a .npmrc pointing to a private registry and only a private registry.  When I "npm install" npm updates every entry in the package.lock file to the public npm registry except for the packages only found in our artifactory server.  I have double and tripple checked the configs.

**Comment by unional** [comment]
I have the exact same problem as @babeal 
We need to lock in the packages to our private registry (artifactory) and after `npm install` some transitive dependencies is resolved to the public `npmjs.org`.

This creates a huge issue for us as we have to keep everything private to ensure build stability.

**Comment by yyfearth** [comment]
The problem I met is some of the dependencies in `package-lock.json` keeping change from `bundled: true` to `resolved: https://npmjs.org/...`
While we are using a private registry configured in .npmrc, and we do not want any dependencies fetch from npmjs.org since we want to npm install can able to work from an env disconnected from Internet.

And the `fsevents` for macos, has huge numbers of bundled dependencies, and keeping doing that, which is very anonying.

**Comment by Justin-ZS** [comment]
Same problem here.
Just run `npm i` without changing `package.json`. 
Some dependencies in `package-lock.json` keeping change from `resolved: https://private.server.here` to `resolved: https://registry.npmjs.org/...`.
But other dependencies still stay at local `resolved`.
Very strange.

**Comment by mattmazzola** [comment]
I agree it's confusing.   ~These are the steps I used to get it to update although it involved regenerating package-lock.json which defeats purpose of package-lock.json.~  I ended up manually editing the package-lock.json to preserve all the previous package versions and minimize risk although the following would have worked if you are not concerned about the packages changing.

1. Delete .npmrc
2. Delete node_modules
3. Delete package-lock.json
4. Run `npm cache clean -f`
5. Run `npm i`

People are expecting to only have to do steps 1 and 5.  I think the mental model is that the installer should detect that registry has changed and update the resolved field on packge-lock.json for them even if contents of package is identical.

**Comment by davidlygagnon** [comment]
I'm also seeing this issue - similar to @Justin-ZS . In my case, I want to point to `https://registry.npmjs.org/` so I added a .npmrc config file with `registry=https://registry.npmjs.org/` and for some reason, not all `resolved` links get updated with the new registry url.

I've done this in the past and it used to work. Maybe related to a specific node version ? I'm on node v8.11.1 and npm 5.6.0

**Comment by davidlygagnon** [comment]
@garygong - that's the issue I was mentioning.

**Comment by mattmazzola** [comment]
I think the biggest question I have is how the hell NPM can still be aware of the previous registry url? I have removed it from all locations if know of on my computer and ran `npm cache clean -f` and it will still populate resolved field with that old registry url even though `npm config get registry` will return the correct default of registry.npmjs.org.  I assume it's in some cache file, but then if that's the case why is this not cleared and how do I manually clear it?

**Comment by rhythmnewt** [comment]
I had the private registry being reverted to the public npm registry consistently when I did "npm i" in my mac environment. Following steps provided by @mattmazzola to clean the cache solved my issue even though I did have to delete the packages-lock.json file (this was acceptable in my case).

What was curious though is that as I was debugging this problem I had some fake authentication in the .npmrc to see if it will trigger any error message but it did not. However after the cache clean, after restoring the .npmrc, I got the auth error (as i would expect) on the first "npm i".

**Comment by cm0s** [comment]
Same issue here
